src/ray/gcs/gcs_server/gcs_kv_manager.h

```diff
@@ -14,43 +14,136 @@
+/// right now.
+class InternalKVInterface {
+ public:
+  /// Get the value associated with `key`
+  ///
+  /// \param key The key to fetch
+  /// \param cb Callback function.
+  virtual void Get(const std::string &key,
+                   std::function<void(std::optional<std::string>)> cb) = 0;
+
+  /// Associate key with value.
+  ///
+  /// \param key The key for the pair
+  /// \param value The value for the pair
+  /// \param overwrite If the key has existed before this op, nothing will happen
+  ///    unless `overwrite` is set to be true.
+  /// \param cb Callback function.
+  virtual void Put(const std::string &key, const std::string &value, bool overwrite,
+                   std::function<void(bool)> cb) = 0;
+
+  /// Delete the key from the store
+  ///
+  /// \param key The key to be deleted
+  /// \param cb Callback function.
+  virtual void Del(const std::string &key, std::function<void(bool)> cb) = 0;
+
+  /// Check whether the key exists in the store
+  ///
+  /// \param key The key to be checked.
+  /// \param cb Callback function.
+  virtual void Exists(const std::string &key, std::function<void(bool)> cb) = 0;
+
+  /// Get the keys for a given prefix.
+  ///
+  /// \param prefix The prefix to be scaned.
+  /// \param cb Callback function.
+  virtual void Keys(const std::string &prefix,
+                    std::function<void(std::vector<std::string>)> cb) = 0;
+
+  virtual ~InternalKVInterface(){};
+};
+
+class RedisInternalKV : public InternalKVInterface {
+ public:
+  RedisInternalKV(RedisClient *redis_client) : redis_client_(redis_client) {}
+
+  void Get(const std::string &key,
+           std::function<void(std::optional<std::string>)> cb) override;
+
+  void Put(const std::string &key, const std::string &value, bool overwrite,
+           std::function<void(bool)> cb) override;
+
+  void Del(const std::string &key, std::function<void(bool)> cb) override;
+
+  void Exists(const std::string &key, std::function<void(bool)> cb) override;
+
+  void Keys(const std::string &prefix,
+            std::function<void(std::vector<std::string>)> cb) override;
+
+ private:
+  RedisClient *redis_client_;
+};
+
+class MemoryInternalKV : public InternalKVInterface {
+ public:
+  MemoryInternalKV(instrumented_io_context *io_context) : io_context_(io_context) {
+    RAY_CHECK(io_context != nullptr);
+  }
+  void Get(const std::string &key,
+           std::function<void(std::optional<std::string>)> cb) override;
+
+  void Put(const std::string &key, const std::string &value, bool overwrite,
+           std::function<void(bool)> cb) override;
+
+  void Del(const std::string &key, std::function<void(bool)> cb) override;
+
+  void Exists(const std::string &key, std::function<void(bool)> cb) override;
+
+  void Keys(const std::string &prefix,
+            std::function<void(std::vector<std::string>)> cb) override;
+
+ private:
+  void RunCB(std::function<void()> cb) { io_context_->post(std::move(cb)); }
```

Comment:
Consider deleting this function and calling io_context_.post() directly for clarity.